src/dict.c

```diff
@@ -534,6 +545,70 @@ void *dictFetchValue(dict *d, const void *key) {
     return he ? dictGetVal(he) : NULL;
 }
 
+void dictSetKey(dict *d, dictEntry* de, void *key) {
+    if (d->type->keyDup)
+        de->key = d->type->keyDup(d, key);
+    else
+        de->key = key;
+}
```

Comment:
it could be that now that that's not a macro, we lose some performance benefits.
let's make some simple benchmark to try to rule that out.

Reply:
Good point. Perhaps with `-O3` this will be inlined? We could check the assembly...

Reply:
We compile with `-flto` (link-time optimization), which turns on optimizations such as inlining across compilation units, so this function is most likely inlined.

I tried adding an assert(0) in `if` to see the stack trace. The function `dictSetKey` was not part of the stack trace, meaning it was inlined.

Reply:
Yeah, I don't think we should try to micro optimize this. We did some brief internal testing at AWS, and found that moving away from #defining sometimes improved performance as we got better branch prediction. Let's just trust the compiler.

Reply:
IMHO, adding exlpicit inline keyword like `inline void dictSetNext()` to those short functions should do good to performance. 

Shall we do as this or just make `inline` a compile option?

Reply:
Fine with me. So far, the only inline we have are `static inline`. Not sure how much the compilers care about inline. Does anyone know?

Reply:
I will stand by my earlier statement, let's just trust the compiler to be doing inlining. We have O3 and flto enabled.

Reply:
@madolson If we move them to the header file, we can't make the struct opaque.

Reply:
Excellent point, sort of defeats the purpose of the whole endeavor :)

Reply:
i agree we want to leave that to the compiler.. just thought it's a good idea to make a minimal effort to benchmark that, so that we don't find a regression later on and start looking for the cause.
hopefully no action will be needed, but if there is, we can discuss it.